x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -151,239 +132,156 @@ public void providersChanged() {
-                        effectiveDescriptors,
-                        roleNames
-                    )
-                );
-                buildThenMaybeCacheRole(
-                    roleKey,
-                    effectiveDescriptors,
-                    rolesRetrievalResult.getMissingRoles(),
-                    rolesRetrievalResult.isSuccess(),
-                    invalidationCounter,
-                    roleActionListener
-                );
-            }, roleActionListener::onFailure));
-        }
+            } else {
+                roleActionListener.onResponse(new Tuple<>(role, role));
+            }
+        }, roleActionListener::onFailure));
     }
 
-    void logDeprecatedRoles(Set<RoleDescriptor> roleDescriptors) {
-        roleDescriptors.stream()
-            .filter(rd -> Boolean.TRUE.equals(rd.getMetadata().get(MetadataUtils.DEPRECATED_METADATA_KEY)))
-            .forEach(rd -> {
-                String reason = Objects.toString(
-                    rd.getMetadata().get(MetadataUtils.DEPRECATED_REASON_METADATA_KEY),
-                    "Please check the documentation"
-                );
-                deprecationLogger.warn(
-                    DeprecationCategory.SECURITY,
-                    "deprecated_role-" + rd.getName(),
-                    "The role [" + rd.getName() + "] is deprecated and will be removed in a future version of Elasticsearch. " + reason
-                );
-            });
-    }
+    public void getRole(Subject subject, ActionListener<Role> roleActionListener) {
+        final Role internalUserRole = tryGetRoleForInternalUser(subject);
+        if (internalUserRole != null) {
+            roleActionListener.onResponse(internalUserRole);
+            return;
+        }
 
-    // for testing
-    Role getXpackUserRole() {
-        return xpackUserRole;
-    }
+        assert false == User.isInternal(subject.getUser()) : "Internal user should not pass here";
 
-    // for testing
-    Role getAsyncSearchUserRole() {
-        return asyncSearchUserRole;
+        final List<RoleReference> roleReferences = subject.getRoleReferences(anonymousUser);
+        // TODO: Two levels of nesting can be relaxed in future
+        assert roleReferences.size() <= 2 : "only support up to one level of limiting";
+
+        buildRoleFromRoleReference(roleReferences.get(0), ActionListener.wrap(role -> {
+            if (roleReferences.size() == 1) {
+                roleActionListener.onResponse(role);
+            } else {
+                buildRoleFromRoleReference(
+                    roleReferences.get(1),
+                    ActionListener.wrap(
+                        limitedByRole -> roleActionListener.onResponse(LimitedRole.createLimitedRole(role, limitedByRole)),
+                        roleActionListener::onFailure
+                    )
+                );
+            }
+
+        }, roleActionListener::onFailure));
```

Comment:
We only support 2 levels of role nesting right now because how `LimitedRole` works. These code can be generalised to support multiple levels once we re-work how Role nesting is represented (likely need an interface for `Role`)